arrows/core/generate_texture.h

```diff
@@ -0,0 +1,441 @@
+  for (unsigned int f = 0; f < mesh->num_faces(); ++f)
+  {
+    vital::matrix_3x3d points_2d_h;
+    points_2d_h << tcoords[f * 3 + 0], tcoords[f * 3 + 1], tcoords[f * 3 + 2], 1, 1, 1;
+    double area_2d = points_2d_h.determinant();
+    auto const& v1 = vertices[triangles(f, 0)];
+    auto const& v2 = vertices[triangles(f, 1)];
+    auto const& v3 = vertices[triangles(f, 2)];
+    vital::vector_3d a3 = v2 - v1;
+    vital::vector_3d b3 = v3 - v1;
+    double area_3d = a3.cross(b3).norm();
+
+    if (area_2d > 0 && area_3d > 0 && !std::isinf(area_2d) && !std::isinf(area_3d))
+    {
+      scale = static_cast<size_t>(std::ceil(sqrt(area_3d / area_2d) / resolution));
+      break;
+    }
+  }
+  for (auto& tc : tcoords)
+  {
+    tc.y() = 1.0 - tc.y();
+    tc *= scale;
+  }
+
+  // Render the depth maps of the mesh seen by the different cameras
+  std::vector<vital::image> depth_maps(images.size());
+  for (unsigned int i = 0; i < images.size(); ++i)
+  {
+    depth_maps[i] = render_mesh_depth_map(mesh, cameras[i])->get_image();
+  }
+
+  // Compute the depth of each points w.r.t each camera
+  std::vector< std::vector<double> > per_camera_point_depth(mesh->num_verts(), std::vector<double>(cameras.size(), 0.0));
+  for (unsigned int v = 0; v < mesh->num_verts(); ++v)
+  {
+    for (unsigned int c = 0; c < cameras.size(); ++c)
+    {
+      per_camera_point_depth[v][c] = cameras[c]->depth(vertices[v]);
+    }
+  }
+
+  std::vector<vital::camera_sptr> cameras_base(cameras.size());
+  for (unsigned int k = 0; k < cameras.size(); ++k)
+  {
+    cameras_base[k] = cameras[k];
+  }
+
+  vital::image_of<T> texture(scale, scale, images[0].depth());
+  vital::transform_image(texture, [](T){ return 0; });
+  kwiver::vital::image_of<char> texture_mask(scale, scale, 1);
+  vital::transform_image(texture_mask, [](char){ return 0; });
+  for (unsigned int f = 0; f < mesh->num_faces(); ++f)
+  {
+    unsigned int p1 = triangles(f, 0);
+    unsigned int p2 = triangles(f, 1);
+    unsigned int p3 = triangles(f, 2);
+    vital::vector_3d const& pt_0 = vertices[p1];
+    vital::vector_3d const& pt_1 = vertices[p2];
+    vital::vector_3d const& pt_2 = vertices[p3];
+
+    render_triangle_from_images<T>(tcoords[f * 3], tcoords[f * 3 + 1], tcoords[f * 3 + 2],
+                                  pt_0, pt_1, pt_2, cameras_base, images,
+                                  per_camera_point_depth[p1],
+                                  per_camera_point_depth[p2],
+                                  per_camera_point_depth[p3],
+                                  depth_maps, texture, 0.1, select_max_score());
+    kwiver::arrows::core::render_triangle<char>(tcoords[f * 3], tcoords[f * 3 + 1],
+                                                tcoords[f * 3 + 2], 1, texture_mask);
+  }
+
+  kwiver::arrows::core::dilate_atlas<T>(texture, texture_mask, 4);
+
+  // Update texture coordinates
+  for (auto& tc : tcoords)
+  {
+    tc[0] += 0.5;   // half-pixel shift
```

Comment:
should we be accounting for this offset in the creation of the texture image instead?  Specifically, in the bilinear interpolation?

Reply:
We could add this offset in the creation of the texture, but I think it is not directly related to the bilinear interpolation of the source image.

It is more related to the texture and how opengl interpolates it. For opengl (0, 0) is the bottom-left corner of the bottom-left pixel and (1, 1) is the top-right corner of the top-right pixel.


If we don't want to do this here, I guess we would have to modify the triangle_scan_iterator (i.e which pixels of the texture are we filling when we render a certain triangle in texture coordinates)